### PR TITLE
Add top-k support to LogitsProcessor

### PR DESCRIPTION
**Pull request by EricLBuehler**
Hello everybody,

During my recent development of `candle-vllm`, I have been working on sampling techniques and noticed that top-k, beam search and some other features are missing from `LogitsProcessor` (#1263).

In this PR, I have added the `top-k` technique (which selects the `k` most likely tokens and samples among those, ensuring that lesser likely tokens are not considered) to `LogitsProcessor`. Additionally, I restructured the `LogitsProcessor` API slightly to use enums in order to reduce the chance for subtle bugs.

### CODE REVIEW

**Comment by LaurentMazare** [comment]
I would suggest you have this enhanced `LogitsProcessor` as part of your `candle-vllm` crate for now - the immediate goal for candle-transformers is not to support all sampling method but instead it would be better if this can be deffered to other crates that can focus on such specific aspects - an alternative would be to make a crate that is only for logits processing if it could be useful without requiring all the dependencies of your crate.

**Comment by EricLBuehler** [comment]
Ok, with that in mind, I think I will make a separate crate for only sampling. Do you think that eventually `candle-transformers`' goal will be to support all sampling methods - perhaps I can reopen an equivalent PR then?
